File: core/src/main/java/com/emarsys/core/request/model/CompositeRequestModel.java
package com.emarsys.core.request.model;

import com.emarsys.core.provider.timestamp.TimestampProvider;
import com.emarsys.core.provider.uuid.UUIDProvider;
import com.emarsys.core.util.Assert;

import java.util.Arrays;
import java.util.Map;

public class CompositeRequestModel extends RequestModel {

    private final String[] originalRequestIds;

    public CompositeRequestModel(
            String id,
            String url,
            RequestMethod method,
            Map<String, Object> payload,
            Map<String, String> headers,
            long timestamp,
            long ttl,
            String[] originalRequestIds) {
        super(url, method, payload, headers, timestamp, ttl, id);
        this.originalRequestIds = originalRequestIds;
    }

    public String[] getOriginalRequestIds() {
        return originalRequestIds;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;

        CompositeRequestModel that = (CompositeRequestModel) o;

        return Arrays.equals(originalRequestIds, that.originalRequestIds);
    }

    @Override
    public int hashCode() {
        int result = super.hashCode();
        result = 31 * result + Arrays.hashCode(originalRequestIds);
        return result;
    }

    @Override
    public String toString() {
        return "CompositeRequestModel{" +
                "request=" + super.toString() +
                "originalRequestIds=" + Arrays.toString(originalRequestIds) +
                '}';
    }

    public static class Builder extends RequestModel.Builder {
        private String[] originalRequestIds;

        public Builder(TimestampProvider timestampProvider, UUIDProvider uuidProvider) {
            super(timestampProvider, uuidProvider);
        }

        public Builder(RequestModel requestModel) {
            super(requestModel);
            Assert.notNull(requestModel, "RequestModel must not be null!");
            this.url = requestModel.getUrl().toString();
            this.headers = requestModel.getHeaders();
            this.method = requestModel.getMethod();
            this.originalRequestIds = ((CompositeRequestModel) requestModel).getOriginalRequestIds();
            this.payload = requestModel.getPayload();
            this.timestamp = requestModel.getTimestamp();
            this.ttl = requestModel.getTtl();
        }

        @Override
        public CompositeRequestModel.Builder url(String url) {
            super.url(url);
            return this;
        }

        @Override
        public CompositeRequestModel.Builder queryParams(Map<String, String> queryParams) {
            super.queryParams(queryParams);
            return this;
        }

        @Override
        public CompositeRequestModel.Builder method(RequestMethod method) {
            super.method(method);
            return this;
        }

        @Override
        public CompositeRequestModel.Builder payload(Map<String, Object> payload) {
            super.payload(payload);
            return this;
        }

        @Override
        public CompositeRequestModel.Builder headers(Map<String, String> headers) {
            super.headers(headers);
            return this;
        }

        @Override
        public CompositeRequestModel.Builder ttl(long ttl) {
            super.ttl(ttl);
            return this;
        }

        public CompositeRequestModel.Builder originalRequestIds(String[] originalRequestIds) {
            this.originalRequestIds = originalRequestIds;
            return this;
        }

        public CompositeRequestModel build() {
            return new CompositeRequestModel(id, buildUrl(), method, payload, headers, timestamp, ttl, originalRequestIds);
        }
    }
}

File: core/src/main/java/com/emarsys/core/request/model/CompositeRequestModel.kt
package com.emarsys.core.request.model

import com.emarsys.core.provider.timestamp.TimestampProvider
import com.emarsys.core.provider.uuid.UUIDProvider


class CompositeRequestModel(
        id: String,
        url: String,
        method: RequestMethod,
        payload: Map<String, Any?>?,
        headers: Map<String, String>,
        timestamp: Long,
        ttl: Long,
        val originalRequestIds: Array<String>) : RequestModel(url, method, payload, headers, timestamp, ttl, id) {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        if (!super.equals(other)) return false
        val that = other as CompositeRequestModel
        return originalRequestIds.contentEquals(that.originalRequestIds)
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + originalRequestIds.contentHashCode()
        return result
    }

    override fun toString(): String {
        return "CompositeRequestModel{" +
                "request=" + super.toString() +
                "originalRequestIds=" + originalRequestIds.contentToString() +
                '}'
    }

    class Builder : RequestModel.Builder {
        private var originalRequestIds: Array<String> = arrayOf()

        constructor(timestampProvider: TimestampProvider, uuidProvider: UUIDProvider) :
                super(timestampProvider, uuidProvider)

        constructor(requestModel: RequestModel) : super(requestModel) {
            originalRequestIds = (requestModel as CompositeRequestModel).originalRequestIds
        }

        override fun url(url: String): Builder {
            super.url(url)
            return this
        }

        override fun queryParams(queryParams: Map<String, String>): Builder {
            super.queryParams(queryParams)
            return this
        }

        override fun method(method: RequestMethod): Builder {
            super.method(method)
            return this
        }

        override fun payload(payload: Map<String, Any?>): Builder {
            super.payload(payload)
            return this
        }

        override fun headers(headers: Map<String, String>): Builder {
            super.headers(headers)
            return this
        }

        override fun ttl(ttl: Long): Builder {
            super.ttl(ttl)
            return this
        }

        fun originalRequestIds(originalRequestIds: Array<String>): Builder {
            this.originalRequestIds = originalRequestIds
            return this
        }

        override fun build(): CompositeRequestModel {
            return CompositeRequestModel(id, buildUrl(), method, payload, headers, timestamp, ttl, originalRequestIds)
        }
    }

}